php_phongo.c

```diff
@@ -1461,6 +1461,45 @@ static const char* php_phongo_bson_type_to_string(bson_type_t type) /* {{{ */
 		bson_iter_key(&(iter)),                        \
 		php_phongo_bson_type_to_string(bson_iter_type(&(iter))))
 
+static bool php_phongo_uri_finalize_auth(mongoc_uri_t* uri TSRMLS_DC) /* {{{ */
+{
+	const char *source = NULL;
+
+	source = mongoc_uri_get_auth_source(uri);
+
+	/* authSource with GSSAPI or X509 should always be external */
+	if (mongoc_uri_get_auth_mechanism(uri)) {
+		if (!strcasecmp(mongoc_uri_get_auth_mechanism(uri), "GSSAPI") ||
+			!strcasecmp(mongoc_uri_get_auth_mechanism(uri), "MONGODB-X509")) {
+			if (source) {
+				if (strcasecmp(source, "$external")) {
+					phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Failed to parse URI options: GSSAPI and X509 require \"$external\" authSource.");
+					return false;
+				}
+			}
```

Comment:
`mongoc_uri_finalize_auth()` also had logic to set the `authSource` to "$external" if it's null at this point. I think we may want to do that as well, as it's possible that the application only specifies `authMechanism` via the URI options array. In that case, we're going to call `mongoc_uri_set_auth_mechanism()` from `php_phongo_apply_options_to_uri()` but leave `authSource` unset.

Looking at the various auth mechanism functions in [mongoc-cluster.c](https://github.com/mongodb/mongo-c-driver/blob/1.15.0/src/libmongoc/src/mongoc/mongoc-cluster.c), I don't think libmongoc is too concerned about this. For instance, the X509 function uses "$external" as a literal instead of actually reading the auth source option; however, it would ensure we leave the URI options in a consistent state regardless of whether the application sets these via query string or array options.

Reply:
Good catch, I added the corresponding else case. Not sure why I removed that in the first place.